sdk/helper/testcluster/docker/environment.go

```diff
@@ -0,0 +1,963 @@
+		"raft": raftOpts,
+	}
+	if opts != nil && opts.VaultNodeConfig != nil && len(opts.VaultNodeConfig.StorageOptions) > 0 {
+		for k, v := range opts.VaultNodeConfig.StorageOptions {
+			if _, ok := raftOpts[k].(string); !ok {
+				raftOpts[k] = v
+			}
+		}
+	}
+	//// disable_mlock is required for working in the Docker environment with
+	//// custom plugins
+	vaultCfg["disable_mlock"] = true
+	vaultCfg["api_addr"] = `https://{{- GetAllInterfaces | exclude "flags" "loopback" | attr "address" -}}:8200`
+	vaultCfg["cluster_addr"] = `https://{{- GetAllInterfaces | exclude "flags" "loopback" | attr "address" -}}:8201`
+
+	systemJSON, err := json.Marshal(vaultCfg)
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(filepath.Join(n.WorkDir, "system.json"), systemJSON, 0o644)
+	if err != nil {
+		return err
+	}
+
+	if opts.VaultNodeConfig != nil {
+		localCfg := *opts.VaultNodeConfig
+		if opts.VaultNodeConfig.LicensePath != "" {
+			b, err := os.ReadFile(opts.VaultNodeConfig.LicensePath)
+			if err != nil || len(b) == 0 {
+				return fmt.Errorf("unable to read LicensePath at %q: %w", opts.VaultNodeConfig.LicensePath, err)
+			}
+			localCfg.LicensePath = "/vault/config/license"
+			dest := filepath.Join(n.WorkDir, "license")
+			err = os.WriteFile(dest, b, 0o644)
+			if err != nil {
+				return fmt.Errorf("error writing license to %q: %w", dest, err)
+			}
+
+		}
+		userJSON, err := json.Marshal(localCfg)
+		if err != nil {
+			return err
+		}
+		err = os.WriteFile(filepath.Join(n.WorkDir, "user.json"), userJSON, 0o644)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create a temporary cert so vault will start up
+	err = n.setupCert("127.0.0.1")
+	if err != nil {
+		return err
+	}
+
+	// setup plugin bin copy if needed
+	copyFromTo := map[string]string{
+		n.WorkDir: "/vault/config",
+		caDir:     "/usr/local/share/ca-certificates/",
+	}
+
+	repo := opts.ImageRepo
+	if repo == "" {
+		repo = "vault"
+	}
+	tag := opts.ImageTag
+	if tag == "" {
+		tag = "latest"
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	var seenLogs uberAtomic.Bool
+	logConsumer := func(s string) {
+		if seenLogs.CAS(false, true) {
+			wg.Done()
+		}
+		n.Logger.Trace(s)
+	}
+	logStdout := &LogConsumerWriter{logConsumer}
+	logStderr := &LogConsumerWriter{func(s string) {
+		if seenLogs.CAS(false, true) {
+			wg.Done()
+		}
+		testcluster.JSONLogNoTimestamp(n.Logger, s)
+	}}
+	r, err := dockhelper.NewServiceRunner(dockhelper.RunOptions{
+		ImageRepo: repo,
+		ImageTag:  tag,
+		// We don't need to run update-ca-certificates in the container, because
+		// we're providing the CA in the raft join call, and otherwise Vault
+		// servers don't talk to one another on the API port.
+		Cmd: []string{"server"},
+		Env: []string{
+			// For now we're using disable_mlock, because this is for testing
+			// anyway, and because it prevents us using external plugins.
+			"SKIP_SETCAP=true",
+			"VAULT_LOG_FORMAT=json",
+		},
+		Ports:           []string{"8200/tcp", "8201/tcp"},
```

Comment:
It might be cool in a future PR to add the ability to use a non-TLS enabled port here.

OCSP and CRL fetching should (theoretically) use a non-TLS listener and I've occasionally had some issues with cert auth method talking to localhost over TLS.

I don't think I need this immediately, perhaps I could add this in the future when I do though. But curious  if you have thoughts on how this would look, it doesn't look like we get that much influence over listeners but I might've missed something.

Reply:
I haven't yet added any support for configuring listeners in HCL, but I don't recall any major obstacles to doing so, just didn't need it yet.